src/ImageCore.jl

```diff
@@ -30,6 +29,16 @@ Color1Array{C<:Color1,N} = AbstractArray{C,N}
 const RRArray{To,From,N,M,P} = Base.ReshapedArray{To,N,Base.ReinterpretArray{To,M,From,P}}
 const RGArray = Union{Base.ReinterpretArray{<:AbstractGray,M,<:Number,P}, Base.ReinterpretArray{<:Number,M,<:AbstractGray,P}} where {M,P}
 
+# delibrately not export these constants to enable extensibility for downstream packages
+const NumberLike{T<:Number} = Union{T,AbstractGray{T}}
+const RealLike{T<:Real} = NumberLike{T}
+const FloatLike{T<:AbstractFloat} = RealLike{T}
+const FractionalLike{T<:Union{FixedPoint, AbstractFloat}} = RealLike{T}
+const IntegerLike{T<:Union{FixedPoint, Integer}} = RealLike{T}
+
+const GrayLike{T<:Union{Bool, FixedPoint, AbstractFloat}} = RealLike{T}
+const GrayImageLike{BT, T<:GrayLike{BT}} = AbstractMatrix{<:RealLike{T}}
```

Comment:
Images are not only 2d, e.g., `testimage("mri")`. It would be fine to have a 2d image type, but put 2d in the name (e.g., `Gray2dImageType`).

What's the purpose of having two type parameters here? Are you hoping to dispatch on `BT`?

Reply:
What's your preference on `Gray2dImageType` vs `Gray2dImage`?

Reply:
> Are you hoping to dispatch on BT?

Yes, occasionally we can do optimization by treating them differently
```julia
foo(img::Gray2dImage)
foo(img::Gray2dImage{FixedPoint})
foo(img::Gray2dImage{AbstractFloat})
```

Reply:
I like `Gray2dImage`.